### PR TITLE
Add optional progress bar when generating build.ninja

### PR DESCRIPTION
**Pull request by nirbheek**
This is just something fun and small I wrote because people were complaining that meson took too long with GStreamer's thousands of targets on Windows.

commit e075ddca4af18ec90ddf5d6696679dd207abe49e:

```
ninja backend: Add optional progress bar in generate()

With, f.ex., gst-build, this step can take 8-20 seconds depending on
your CPU and disk speed. I was looking at speeding that up, but the
ways I could see would've involved heavy refactoring of the backend
generation, and would not have solved the core problem of no feedback.

Now we have pretty (optional) progress bars!
```

commit b1600771d90cf267503402d0fd57faad921a0532:

```
compilers: Clarify a docstring in CompilerArgs
```

### CODE REVIEW

**Comment by jpakkane** [comment]
meson.build:3:0: ERROR: Executables created by cpp compiler /usr/bin/x86_64-w64-mingw32-g++ are not runnable.

Most likely unrelated?

As for the code, we already have one progress bar printer (in wrap file download code). Could we make both use the same implementation? Though it might be tricky since the download code works on bytes, not via a generator like tqdm does.

**Comment by nirbheek** [comment]
> Could we make both use the same implementation? Though it might be tricky since the download code works on bytes, not via a generator like tqdm does.

Looking at the code, that sounds doable, tqdm supports manual updates too. I'll keep the existing code for the case when tqdm is not installed and make it use tqdm when it is :)

File: mesonbuild/mesonlib.py [hunk 1]
```diff
@@ -1391,3 +1391,58 @@ class LibType(Enum):
...
+            return
+        ratio = int(self.done / self.total * 10)
+        while self.printed_dots < ratio:
+            self.print_dot()
+
+    def close(self):
+        print('')
+
+try:
+    raise ImportError
```
Comment (jpakkane) [inline mesonbuild/mesonlib.py]: Forgotten debug code?

If yes and this needs fixing then also consider switching the names around? That is, provide and use a `ProgressBar` class in our code that may (or may not) be implemented with `tqdm`. The latter is a confusing name to encounter out of the blue.

Comment (nirbheek) [reply mesonbuild/mesonlib.py]: Yes, that's debug code. Good point re: implementing it via the class. I'll make it happen.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

**Comment by nirbheek** [comment]
Updated.

**Pull request merged**